src/operation.ts

```diff
@@ -0,0 +1,42 @@
+// operation.ts
+
+/*
+ * Copyright (c) 2021-2023 Check Digit, LLC
+ *
+ * This code is licensed under the MIT license (see LICENSE.txt for details).
+ */
+
+export interface CreateOperation<Item> {
+  type: 'create';
+  createdOn: string;
+  item: Item;
+}
+
+export interface DeleteOperation {
+  type: 'delete';
+  createdOn: string;
+  name: string;
+  previousCreatedOn: string;
+}
+
+export type Operation<Item> = CreateOperation<Item> | DeleteOperation;
+
+export function getItemsFromOperations<Item extends { name: string }>(
+  operations: Operation<Item>[],
+  at: string,
+): Item[] {
+  const itemMap = new Map<string, Item>();
+
+  const sortedOperations = operations.sort((operation1, operation2) =>
+    operation1.createdOn.localeCompare(operation2.createdOn),
+  );
+  for (const operation of sortedOperations) {
+    if (operation.type === 'create' && operation.createdOn < at) {
+      itemMap.set(operation.item.name, operation.item);
+    } else if (operation.type === 'delete' && operation.createdOn < at) {
+      itemMap.delete(operation.name);
```

Comment:
i just noticed this but doesn't this need to take the previousCreatedOn into account when performing the deletion?

Reply:
Given the scenario `{
    type: 'create',
    item: { name: 'xxxxxx, alphabeticCode: 'yyy', numericCode: 'zzz', minorUnits: 2 },
    createdOn: '1970-01-01T00:00:00.000Z',
  },
  {
    type: 'delete',
    name: 'xxxxxx',
    previousCreatedOn: '1970-01-01T00:00:00.000Z',
    createdOn: '2023-04-15T00:00:00.000Z',
  },
  {
    type: 'create',
    item: { name: 'xxxxxx', alphabeticCode: 'yyy', numericCode: 'zzz', minorUnits: 0 },
    createdOn: '2023-04-15T00:00:00.000Z',
  },`  if we pass at as '2023-04-15T00:00:00.000Z', it should return     item: { name: 'xxxxxx, alphabeticCode: 'yyy', numericCode: 'zzz', minorUnits: 2 },
but if we go by previousCreatedOn on deletion it would pick the above record and would finally delete same record and would end up throwing an error.

Reply:
by utilizing previousCreatedOn I don't mean using that as a match against the at value. i mean when the itemMap.delete(operation.name) is done, you want to match not only on name but on createdOn === previousCreatedOn. because everything is sorted by createdOn i'm not sure if it's necessary but if it's not, then we shouldn't even need the previouslyCreatedOn value that's being included with the deletion.

Reply:
utilizing the previouslyCreatedOn value  seems to be efficient compared to sorting.